src/borg/archiver.py

```diff
@@ -696,6 +698,219 @@ def peek_and_store_hardlink_masters(item, matched):
+            # atime, ctime, possibly Linux capabilities (security.* xattrs)
+            # and various additions supported by GNU tar in POSIX mode.
+
+            stream = None
+            tarinfo = tarfile.TarInfo()
+            tarinfo.name = item.path
+            tarinfo.mtime = item.mtime / 1e9
+            tarinfo.mode = stat.S_IMODE(item.mode)
+            tarinfo.uid = item.uid
+            tarinfo.gid = item.gid
+            tarinfo.uname = item.user or ''
+            tarinfo.gname = item.group or ''
+            # The linkname in tar has the same dual use the 'source' attribute of Borg items,
+            # i.e. for symlinks it means the destination, while for hardlinks it refers to the
+            # file.
+            # Since hardlinks in tar have a different type code (LNKTYPE) the format might
+            # support hardlinking arbitrary objects (including symlinks and directories), but
+            # whether implementations actually support that is a whole different question...
+            tarinfo.linkname = ""
+
+            modebits = stat.S_IFMT(item.mode)
+            if modebits == stat.S_IFREG:
+                tarinfo.type = tarfile.REGTYPE
+                if 'source' in item:
+                    source = os.sep.join(item.source.split(os.sep)[strip_components:])
+                    if hardlink_masters is None:
+                        linkname = source
+                    else:
+                        chunks, linkname = hardlink_masters.get(item.source, (None, source))
+                    if linkname:
+                        # Master was already added to the archive, add a hardlink reference to it.
+                        tarinfo.type = tarfile.LNKTYPE
+                        tarinfo.linkname = linkname
+                    elif chunks is not None:
+                        # The item which has the chunks was not put into the tar, therefore
+                        # we do that now and update hardlink_masters to reflect that.
+                        item.chunks = chunks
+                        tarinfo.size = item.get_size()
+                        stream = item_content_stream(item)
+                        hardlink_masters[item.get('source') or original_path] = (None, item.path)
+                else:
+                    tarinfo.size = item.get_size()
+                    stream = item_content_stream(item)
+            elif modebits == stat.S_IFLNK:
+                tarinfo.type = tarfile.SYMTYPE
+                tarinfo.linkname = item.source
+            elif modebits == stat.S_IFBLK:
+                tarinfo.type = tarfile.BLKTYPE
+                tarinfo.devmajor = os.major(item.rdev)
+                tarinfo.devminor = os.minor(item.rdev)
+            elif modebits == stat.S_IFCHR:
+                tarinfo.type = tarfile.CHRTYPE
+                tarinfo.devmajor = os.major(item.rdev)
+                tarinfo.devminor = os.minor(item.rdev)
+            elif modebits == stat.S_IFDIR:
```

Comment:
maybe move this up a bit, below regular files, as this is likely the 2nd most frequent fs object, more frequent than symlinks, blk/chr devices.